### PR TITLE
Resolve WAL delta for shard diff transfer

### PR DESCRIPTION
**Pull request by timvisee**
Tracked in: <https://github.com/qdrant/qdrant/issues/3477>
Depends on: <https://github.com/qdrant/qdrant/pull/3510>, <https://github.com/qdrant/qdrant/pull/3551>

Add a function to resolve a WAL delta from some recovery point.

It is based on what we described in our design document, but might need another pass to get all the fine details right.

This will be used in <https://github.com/qdrant/qdrant/pull/3509> to actually select a part of the WAL for the shard diff transfer.

### Tasks
- [x] Merge <https://github.com/qdrant/qdrant/pull/3510>
- [x] Merge <https://github.com/qdrant/qdrant/pull/3551>
- [x] Rebase on `dev`
- [ ] Resolve open TODOs
- [x] Undraft

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [ ] Does your submission pass tests?
2. [x] Have you formatted your code locally using `cargo +nightly fmt --all` command prior to submission?
3. [x] Have you checked your code using `cargo clippy --all --all-features` command?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: lib/collection/src/shards/local_shard/mod.rs [hunk 1]
```diff
@@ -919,6 +920,61 @@ impl LocalShard {
...
+        // Scroll back over the WAL and find a record that covered all clocks
+        // Drain satisfied clocks from the recovery point until we have nothing left
+        log::trace!("Resolving WAL delta for: {recovery_point:?}");
+        let delta_from = self
+            .wal
+            .lock()
+            .read_from_last(true)
+            .filter_map(|(op_num, update)| update.clock_tag.map(|clock_tag| (op_num, clock_tag)))
+            // Skip forced clock tags
+            .filter(|(_, clock_tag)| !clock_tag.force)
```
Comment (timvisee) [inline lib/collection/src/shards/local_shard/mod.rs:968]: On second thought: we probably want to skip this condition, to allow forced entries.

Comment (timvisee) [reply lib/collection/src/shards/local_shard/mod.rs]: I removed this condition now.

File: lib/collection/src/shards/local_shard/mod.rs [hunk 2]
```diff
@@ -919,6 +920,61 @@ impl LocalShard {
...
+        // Ensure the recovering node gets records for a clock it might not have seen yet
+        recovery_point.extend_with_missing_clocks(&last_seen_snapshot);
+
+        // Remove clocks that are equal to the current last seen
+        // We don't have to transfer any record for these
+        // TODO: do we want to remove higher clocks too, as the recovery node already has all data?
+        recovery_point.remove_equal_clocks(&last_seen_snapshot);
+
+        // TODO: check truncated clock values or each clock we have:
+        // TODO: - if truncated is higher, we cannot resolve diff
```
Comment (timvisee) [inline lib/collection/src/shards/local_shard/mod.rs:957]: We might want to merge with these TODOs in place, so we can resolve them later once a separate truncated clock map is implemented.

File: lib/collection/src/wal.rs [hunk 3]
```diff
@@ -156,6 +156,28 @@ impl<'s, R: DeserializeOwned + Serialize + Debug> SerdeWal<R> {
         })
     }
 
+    /// Read all records from newest to oldest
+    ///
+    /// Normally this only iterates over our logical WAL, meaning all acknowledged records are
+    /// excluded. If `with_physical` is set to `true`, it will also iterate over all acknowledged
+    /// records we still have on disk.
+    pub fn read_from_last(&'s self, with_physical: bool) -> impl Iterator<Item = (u64, R)> + 's {
```
Comment (generall) [inline lib/collection/src/wal.rs:164]: 👊

File: lib/collection/src/shards/local_shard/mod.rs [hunk 4]
```diff
@@ -919,6 +920,59 @@ impl LocalShard {
...
     }
+
+    /// Resolve the WAL delta on the current shard for the given `recovery_point`
+    ///
+    /// The delta can be sent over to the node which the recovery point is from, to restore its
+    /// WAL making it consistent with the current shard.
+    ///
+    /// On success, a WAL record number from which the delta is resolved is returned.
+    /// If a WAL delta could not be resolved, an error is returned describing the failure.
+    pub async fn resolve_wal_delta(
```
Comment (generall) [inline lib/collection/src/shards/local_shard/mod.rs:931]: we might want to decouple this fuction from it's implementation and `self`, cause we might want to test it independently from LocalShard (which requires a lot of hustle to init properly)

File: lib/collection/src/wal.rs (hunk 3 above)
Comment (timvisee) [reply lib/collection/src/wal.rs]: `with_🥊` is now `with_acknowledged`.

**Review by generall — APPROVED** [review]
Let's merge and start tests :+1:

**Pull request merged**